exdir/core/dtype.py

```diff
@@ -0,0 +1,123 @@
+from numpy import dtype
+from .links import Reference, RegionReference
+
+ref_dtype = dtype(str, metadata={'ref': Reference})
+regionref_dtype = dtype(str, metadata={'ref': RegionReference})
+
+
+def special_dtype(**kwds):
+    """ Create a new h5py "special" type.  Only one keyword may be given.
+    Legal keywords are:
+    vlen = basetype
+        Base type for HDF5 variable-length datatype. This can be Python
+        str type or instance of np.dtype.
+        Example: special_dtype( vlen=str )
+    enum = (basetype, values_dict)
+        Create a NumPy representation of an HDF5 enumerated type.  Provide
+        a 2-tuple containing an (integer) base dtype and a dict mapping
+        string names to integer values.
+    ref = Reference | RegionReference
+        Create a NumPy representation of an HDF5 object or region reference
+        type.
+    """
```

Comment:
This code is copied from `h5py`. This file should therefore have a header stating that it originates from `h5py` and has a BSD license. Something similar to their own headers is fine, except it should not state that it is part of `h5py`, of course:

```
# This file is part of h5py, a Python interface to the HDF5 library.
#
# http://www.h5py.org
#
# Copyright 2008-2019 Andrew Collette and contributors
#
# License:  Standard 3-clause BSD; see "license.txt" for full license terms
#           and contributor agreement.
```

In addition, we should add a `third-party-licenses/h5py.txt` file to the repo.

The mentions of HDF5 and h5py in the code itself should probably be replaced with `exdir` to avoid confusion.